working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1081,11 +1049,11 @@ expression.
 ```dart
 main() {
   switch (some(extremely, long, expression, here)) {
-    case Some(Quite(var long, var pattern)) => expression();
+    case Some(Quite(:var long, :var pattern)) => expression();
   };
```

Comment:
That's an interesting point. Lambdas already do too (modulo `await`/`yield`/`break`/`continue`). But, yes, switch expressions give you full expression locals even inside loops, async functions, and generators.

Whether that suggests we should do `let` expressions or some other syntax is a separate discussion. :)

Reply:
We need to be very clear on the scoping of such variables. Each evaluation must introduce a new binding since the variable can be final. That's *probably* ensured by putting the variable in the nearest enclosing scope, because there is no way to evaluate the same expression twice (aka. loop) without introducing a new scope per iteration. We just need to make sure we haven't forgotten (or introduced) any operation where that's not true.

For switch cases, we need a scope per `casehead`, because each can introduce independent final variables that can be read in the `when` clause, whether the case is then accepted or not.

Not a problem, we just need to be careful.